<a href="https://colab.research.google.com/github/yewon91/mul_lectures/blob/main/%ED%86%B5%EA%B3%84%EB%B6%84%EC%84%9D_240111.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 문제 상황
- 편의점 감자튀김 (무게 130g으로 알려져 있음)
- 학생이 무게 하나 잼. 122.02g 밖에 되지 않음
- 2주간 매일 감자튀김을 사서 무게 측정
  + 14개의 표본의 평균 계산했는데, 128.451g ==> 학생이 편의점주에게 항의
  + 편의점주 왈 : 그저 우연이다!

### 질문
- 어떤 통계 검정 방법을 사용?
  + t-test, One sample T-test
  + 표본 14개 => 정규성 검정 진행 후, 정규성 검정 통과할 때, One Sample T-Test
- 귀무가설과 대립가설 설정?
  + 귀무가설 : 모평균(130g)과 표본평균(128.451)과 같다
  + 대립가설 : 모평균(130g)과 표본평균(128.451)과 같지 않다 (양측검정)
  + 대립가설 : 모평균(130g) > 표본평균(128.451) 보다 크다 (단측검정)

In [ ]:
!pip install scipy

     ---------------------------------------- 0.0/60.4 kB ? eta -:--:--
     ------------ ------------------------- 20.5/60.4 kB 330.3 kB/s eta 0:00:01
     -------------------------------------- 60.4/60.4 kB 643.1 kB/s eta 0:00:00
   ---------------------------------------- 0.0/44.1 MB ? eta -:--:--
   ---------------------------------------- 0.2/44.1 MB 5.3 MB/s eta 0:00:09
   ---------------------------------------- 0.4/44.1 MB 6.4 MB/s eta 0:00:07
    --------------------------------------- 0.8/44.1 MB 6.6 MB/s eta 0:00:07
   - -------------------------------------- 1.2/44.1 MB 7.1 MB/s eta 0:00:07
   - -------------------------------------- 1.5/44.1 MB 6.7 MB/s eta 0:00:07
   - -------------------------------------- 2.0/44.1 MB 7.3 MB/s eta 0:00:06
   -- ------------------------------------- 2.3/44.1 MB 7.5 MB/s eta 0:00:06
   -- ------------------------------------- 2.8/44.1 MB 7.6 MB/s eta 0:00:06
   -- ------------------------------------- 3.1/44.1 MB 7.7 MB/s eta 0:00:06
   --


[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
import pandas as pd
import numpy as np
from scipy import stats

In [ ]:
%precision 3
np.random.seed(1111)

In [ ]:
import scipy
scipy.__version__

'1.11.4'

In [ ]:
df = pd.read_csv("data/ch11_potato.csv")
sample = np.array(df['무게'])
sample

array([122.02, 131.73, 130.6 , 131.82, 132.05, 126.12, 124.43, 132.89,
       122.79, 129.95, 126.14, 134.45, 127.64, 125.68])

In [ ]:
# 표본평균
s_mean = np.mean(sample)
s_mean

128.451

### A학생의 관심
- (14개의)표본평균이 모평균 130g보다 적은지 여부
- 여기서는 표본평균이 정규분포를 따르고 있고 ==> 정규성 검정을 할 필요가 없다
  + 비모수 검정을 쓸 필요가 없다!!
  + 모수 검정을 하면 된다!!
- 모분산이 9임을 알고 있다고 전제한다

In [ ]:
# 정규분포 객체 생성
rv = stats.norm(130, #모평균
                np.sqrt(9/14)) # 9 모분산, 14 표본의 개수
rv.isf(0.95) # 신뢰구간 95%

128.681

In [ ]:
rv = stats.norm(130, #모평균
                np.sqrt(9/30)) # 9 모분산, 30 표본의 개수
rv.isf(0.95) # 신뢰구간 95%

129.099

# 검정통계량 (=t-통계량)
 - 샘플개수가 변해도 임계값은 동일

In [ ]:
# 130 모평균
z = (s_mean - 130) / np.sqrt(9/14)
z

-1.932

In [ ]:
# 검정통계량에 관한 임계값 구하기
rv = stats.norm()
rv.isf(0.95)

-1.645

In [ ]:
# 130 모평균
z = (s_mean - 130) / np.sqrt(9/30)
rv = stats.norm()
rv.isf(0.95)

-1.645

In [ ]:
rv.cdf(z) # p값, 유의수준 0.05보다 작음

# 귀무가설 채택 or 귀무가설 기각
# 대립가설 채택(X)

0.002

In [ ]:
# 양측검정
z = (s_mean - 130) / np.sqrt(9/14)
rv = stats.norm()
rv.interval(0.95) # 신뢰구간 95%

(-1.960, 1.960)

In [ ]:
rv.cdf(z) * 2 # p-value

0.053

## 양측검정, 단측검정
- 양측검정 A와 B가 다르다(같지 않다)
  + 해석 : 일반적인 평균값 가지고 비교 가능
- 단측검정 A>B, B>A

## 가설검정의 두 가지 오류
- 제 1종 오류 : 귀무가설이 옳을 때 귀무가설을 기각하는 오류
- 제 2종 오류 : 대립가설이 옳을 때 귀무가설을 채택하는 오류

### 제 1종 오류
- 실제 평균이 130g인데도 불구하고 평균은 130g보다 작다고 결론을 내리는 상황
    + False Positive (오탐)

In [ ]:
rv = stats.norm(130, 3)
n_samples = 10000
c = stats.norm().isf(0.95)
cnt = 0
for _ in range(n_samples):
    sample_ = np.round(rv.rvs(14), 2)
    s_mean_ = np.mean(sample_)
    z = (s_mean_ - 130) / np.sqrt(9/14)
    if z < c:
        cnt += 1
cnt / n_samples

0.053

- 1종오류를 범할 비율 0.053 / 약 5%의 비율로 130g보다 작다고 잘못탐지
- 1종오류를 범할 확률: 위험률
- 좀더 엄격하게 적용 => 유의수준 1%에서 가설검정 수행

### 제 2종 오류
- 실제 표본평균이 130g보다 작음에도 불구하고 그런 결론을 얻을 수 없는 상황
  + False negative (미탐)

In [ ]:
rv = stats.norm(128, 3)
c = stats.norm().isf(0.95)
n_samples = 10000
cnt = 0
for _ in range(n_samples):
    sample_ = np.round(rv.rvs(14), 2)
    s_mean_ = np.mean(sample_)
    z = (s_mean_ - 130) / np.sqrt(9/14)
    if z >= c:
        cnt += 1

cnt / n_samples

0.197

- 0.197 의미는 검정력power(1-0.197) / 제대로 탐지못할 확률 20%
- 제2종 오류는 언제나 모집단의 정보에 의존한다 (모평균이 낮아지면 미탐 발생확률 낮아짐)
  + point: 분석가는 모집단의 구체적인 정보를 알 방법이 없다
  + 1종오류, 2종오류는 Trade-off 관계
  + 1종오류만 신경쓰면 됨
- 현업에서는 모분산(모평균)을 알 수 없음 ==> One Sample T-test는 잘 안씀

In [ ]:
rv = stats.norm(127, 3)
c = stats.norm().isf(0.95)
n_samples = 10000
cnt = 0
for _ in range(n_samples):
    sample_ = np.round(rv.rvs(14), 2)
    s_mean_ = np.mean(sample_)
    z = (s_mean_ - 130) / np.sqrt(9/14)
    if z >= c:
        cnt += 1

cnt / n_samples

0.017

### 모분산을 안다
- 현실적으로 모집단을 아는 것은 불가능
  + 건강보험, 통계 / 전국민 데이터가 있는 자료 (국가)
  + 모분산을 안다 => 모집단을 안다 => 모평균도 구할 수 있음 => 모평균에 대한 검정을 우리가 굳이 할 필요가 있나?

In [ ]:
def pmean_test(sample, mean0, alpha=0.05):
    '''
    sample: 샘플 데이터
    mean0: 모평균
    alpha: 95% 수준의 신뢰구간
    '''
    s_mean = np.mean(sample)
    u_var = np.var(sample, ddof=1)
    n = len(sample)
    rv = stats.t(df=n-1)
    interval = rv.interval(1-alpha)

    t = (s_mean - mean0) / np.sqrt(u_var/n)
    print(f't통계량 값은 {t:.3f}')
    if interval[0] <= t <= interval[1]:
        print('귀무가설을 채택')
    else:
        print('귀무가설을 기각')

    if t < 0:
        p = rv.cdf(t) * 2
    else:
        p = (1 - rv.cdf(t)) * 2
    print(f'p값은 {p:.3f}')

In [ ]:
pmean_test(sample, 130)

t통계량 값은 -1.455
귀무가설을 채택
p값은 0.169


In [ ]:
t, p = stats.ttest_1samp(sample, 130)
t, p

(-1.455, 0.169)

### z 통계량 / t 통계량
- 모집단,모평균,모분산을 다 안다.
  + z-검정 : z-통계량

- 표본을 추출한다. 표본평균, 표본 분산을 안다
  + t-검정 : t-통계량

## 비전공자 / 이공계
- 비전 : 해당수식을 일반인들에게 어떻게 쉽게 풀어 설명해줄 것인가 (보고서 작성시)
- 이공계: 수식정리해서 블로그 정리

# 2표본 문제에 관한 가설검정
- 대응표본 : 정규분포 가정O
  + 정규성 검정 통과 : 대응표본 t-검정
  + 정규성 검정 통과 실패 : 비모수 검정, 윌콕슨의 부호순위검정

- 독립표본 : 정규분포 가정O
  + 정규성 검정 통과 : 독립비교 t-검정
  + 정규성 검정 통과 실패 : 맨-위트니의 U 검정

### 대응표본 t-검정
- A학생은 친구 20명에게 1주일간 근력운동을 하게 함(intervention, 개입)
- 운동전후 집중력 측정하는 테스트
- 가설검정
  + 귀무가설 : 운동 전후 집중력 평균에 차이가 없을 것이다
  + 대립가설 : 운동 전후 집중력 평균에 차이가 있을 것이다

In [ ]:
# import pandas as pd

training_rel = pd.read_csv('data/ch11_training_rel.csv')
training_rel.head(1)

전   후
0  59  41

In [ ]:
training_rel['차'] = training_rel['후'] - training_rel['전']

In [ ]:
from scipy import stats
t, p = stats.ttest_1samp(training_rel['차'], 0)
t, p

(2.204, 0.040)

In [ ]:
t, p = stats.ttest_rel(training_rel['후'], training_rel['전'])
t, p

(2.204, 0.040)

## 보고서 작성
- t-통계량 / p-value값

In [ ]:
전,후 시각화코드 / 전,후 평균

SyntaxError: invalid syntax (757740915.py, line 1)

## 독립비교 t-검정
- A,B그룹 집중력 평균 차이 테스트
- 어떤 검정을 수행해야?
- 가설설정
  + 귀무가설: A그룹의 평균과 B그룹의 평균 차이는 없다.
  + 대립가설: A그룹의 평균과 B그룹의 평균 차이는 있다.

- 독립비교 t-검정 두가지 가정
  1) 각 그룹 데이터의 정규성 검정
  2) 두그룹의 분산이 같다 (=등분산성 검정), levene 검정 시행

In [ ]:
training_ind = pd.read_csv('data/ch11_training_ind.csv')
training_ind.shape

- 결론은 두그룹의 평균 차이는 없다, 귀무가설 채택

In [ ]:
t, p = stats.ttest_ind(training_ind['A'], training_ind['B'], equal_var = True)
t, p

In [ ]:
# Levene 검정 이후 살펴봤더니 등분산성이 다르더라
t, p = stats.ttest_ind(training_ind['A'], training_ind['B'], equal_var = False)
t, p

## 비모수 검정
- 모수검정에서 각 가정을 만족시키지 못할때 사용

### 윌콕슨의 부호순위검정
- 대응표본에서 정규분포 만족하지 못할때
- 평균차이(X), 중앙값 차이에 대한 검정(O) - 50%에 해당하는 값

In [ ]:
training_rel = pd.read_csv('data/ch11_training_rel.csv')
toy_df = training_rel[:6].copy()
toy_df['차'] = toy_df['후'] - toy_df['전']
toy_df

In [ ]:
# 순위 적용
diff = toy_df['후'] - toy_df['전']
rank = stats.rankdata(abs(diff)).astype(int)
toy_df['순위'] = rank
toy_df

In [ ]:
import numpy as np

r_minus = np.sum((diff < 0) * rank)
r_plus = np.sum((diff > 0) * rank)

r_minus, r_plus # 검정통계

(diff < 0) * rank

In [ ]:
toy_df['후'] = toy_df['전'] + np.arange(1, 7)
diff = toy_df['후'] - toy_df['전']
rank = stats.rankdata(abs(diff)).astype(int)
toy_df['차'] = diff
toy_df['순위'] = rank
toy_df

r_minus = np.sum((diff < 0) * rank)
r_plus = np.sum((diff > 0) * rank)

r_minus, r_plus # 검정통계